Feature/hermes sharing

jnation3406 opened the pull request:
So this is a rather large PR - I'm happy to go through it with you after the AAS.

Major Changes:
1. Add "open in hermes" links to a few of the sharing methods to preload data into hermes (pending a PR in hermes)
2. Remove the sharing section lower down the left bar of the target details page since it was redundant to other methods
3. Enable target_list sharing in hermes.
4. Attempt to standardize the sharing forms a bit more to all have title/message fields (only used for hermes sharing)

jchate6 commented:
Trying to do the "Open in Hermes" with ~30 Data points got me a `414 Request-URI Too Large` error. Which is a new one for me.

jnation3406 commented:
> Trying to do the "Open in Hermes" with ~30 Data points got me a `414 Request-URI Too Large` error. Which is a new one for me. (pointed to hermes-dev)

Yeah, I've deployed it to hermes-dev but havent updated the nginx yet to accept much larger URIs... its on my list

jchate6 commented on tom_targets/templatetags/targets_extras.py:
```diff
@@ -48,10 +50,24 @@ def target_feature(target):
+    sharing = getattr(settings, "DATA_SHARING", None)
+    hermes_url_with_data = ''
+    hermes_url_no_data = ''
+    if sharing and 'hermes' in sharing:
```
If we go with the above idea to just include this on the sharing page, we can probably remove this here.

jchate6 reviewed (changes requested):
I think these changes are great!
I've made some comments, questions and suggestions. Let me know if you'd like to take care of these or if you'd rather I worked on them myself.

Finally, I think it makes a lot of sense to add some tests to `tom_dataproducts` and `tom_targets` to make sure we are handling the conversions correctly and handling feedback as expected. Obviously we should also have some canary tests to make sure we properly catch any Hermes changes, but that should maybe wait until our canary test overhall.

jnation3406 commented:
> > Trying to do the "Open in Hermes" with ~30 Data points got me a `414 Request-URI Too Large` error. Which is a new one for me. (pointed to hermes-dev)
> 
> Yeah, I've deployed it to hermes-dev but havent updated the nginx yet to accept much larger URIs... its on my list

I've updated hermes-dev so it hopefully will work for larger datasets opening in hermes. I have it like 5Mb for header size, so not sure exactly how large but assume hundreds of datapoints would be fine now.

jnation3406 commented on tom_targets/templates/tom_targets/partials/target_buttons.html:
```diff
@@ -1,8 +1,21 @@
+        <div class="dropdown-menu">
+            {% if "hermes" in sharing.keys %}
+            <h6 class="dropdown-header">Open in Hermes</h6>
+            <a href="{{ hermes_url_no_data }}" title="Hermes Target Only" target="_blank" class="dropdown-item">Target Only &#10064;</a>
```
Agreed, I'll switch it.

jnation3406 commented on tom_dataproducts/alertstreams/hermes.py:
```diff
@@ -128,16 +138,17 @@ def get_hermes_topics(**kwargs):
+            submit_url = stream_base_url + "api/v0/profile/"
+            headers = {'Authorization': f"Token {settings.DATA_SHARING['hermes']['HERMES_API_KEY']}"}
+            response = requests.get(url=submit_url, headers=headers)
+            topics = response.json()['writable_topics']
```
Yeah, I was going to say that the sharing module is where it does the filtering by the `USER_TOPICS`. There I changed the behavior so that it uses `USER_TOPICS` to pair down the list of what is returned by the api if it is specified. We could return an empty list if the profile query fails, but by returning USER_TOPICS we make the issue of losing connection to hermes when getting the profile potentially recoverable (since you can still submit on the TOM, and if you have connection to hermes again and its a writable topic it will suceed). Not sure if its worth it keeping it for that reason, since it is also potentially confusing.

jnation3406 replied:
After reading some of your other comments I agree with returning an empty list if getting the profile fails.

jnation3406 commented on tom_dataproducts/sharing.py:
```diff
@@ -215,17 +245,19 @@ def get_sharing_destination_options():
+            destination_topics = details.get('USER_TOPICS', [])
+            if destination.lower() == 'hermes':
+                # If this is a hermes share, get the topics from hermes and override what the users provide
+                hermes_topics = get_hermes_topics()
```
okay, I guess I'm on board with that + returning an empty topic set if we can't get writable topics. That way they won't even see hermes as an option.

jnation3406 commented on tom_dataproducts/templates/tom_dataproducts/partials/dataproduct_list_for_target.html:
```diff
@@ -46,20 +46,22 @@ <h4>Data</h4>
               {% buttons %}
-                <input type="submit" class="btn btn-primary" value="Submit" name="share_dataproduct_form" style="position:absolute; bottom:0">
+                <input type="submit" class="btn btn-primary" value="Submit" name="share_dataproduct_form" style="position:absolute; bottom:1rem">
               {% endbuttons %}
```
Yeah exactly, Since dataproducts map to actual files that should be sent through hermes and then loaded into another TOM. But unfortunately, there is no way to "open in hermes" with the files pre-selected (since they only exist on your TOM storage) so this sharing will have to just use the form, but could then programatically send the files to hermes.

jnation3406 commented on tom_dataproducts/templates/tom_dataproducts/partials/photometry_datalist_for_target.html:
```diff
@@ -27,7 +27,7 @@
             {% for datum in data %}
             <tr>
-                <td><input type="checkbox" class="phot-row" id="phot-row-{{datum.id}}" name="share-box" value="{{datum.id}}" onchange="check_selected()"></td>
+                <td><input type="checkbox" class="phot-row" id="phot-row-{{ forloop.counter0 }}" name="share-box" value="{{ forloop.counter0 }}" onchange="check_selected({{ forloop.counter0 }})"></td>
```
Ah I see - I will change the `value={{ datum.id }}` back. That way it will work with the share form and with the open in hermes (which is using the id/onchange, not that value).

jnation3406 commented:
Just wanted to update that I've made most of the changes you requested, but in testing I couldn't get our deploy of hermes to work with any reasonable sized preloading from the URI, so I've been  working on another solution for preloading that doesn't depend on large URIs and should work with our deployment infrastructure. I hope to have that deployed along with the updates to this branch in the next day or so.

jnation3406 commented:
I've addressed your original comments, and also refactored a bit to work with the different method of preloading data into the hermes form. This new method is deployed to hermes-dev.lco.global right now, and it doesn't depend on storing things in the URL so it should be safe for any size of data being shared. The only caveat is the data being preloaded is required to pass hermes validation, i.e. your datums must have an instrument or telescope set on them... But this is also a requirement for the data to be shared through hermes through the form anyway.

jchate6 commented on tom_dataproducts/alertstreams/hermes.py:
```diff
@@ -28,43 +29,83 @@ def __init__(self, title='', submitter='', authors='', message='', topic='hermes
+        hermes_alert.save()
+        for tomtoolkit_photometry in datums:
+            tomtoolkit_photometry.message.add(hermes_alert)
+    except Exception:
```
Should this be a more specific exception or list of exceptions? 
This seems like it could hide other errors.

jnation3406 replied:
Hmm I think we probably want to catch any exception so the page doesn't crash. The response is returned so if it's an error response or doesn't contain the message we expect, then we will report the error back to the user.

jchate6 reviewed (changes requested):
just a few more nits...

jnation3406 commented on tom_dataproducts/alertstreams/hermes.py:
```diff
@@ -28,43 +29,83 @@ def __init__(self, title='', submitter='', authors='', message='', topic='hermes
+    return alert
 
 
 def create_hermes_target_table_row(target, **kwargs):
```
Interesting I'll try to get a nonsidereal target in my TOM to test...

jnation3406 commented on tom_dataproducts/alertstreams/hermes.py:
```diff
@@ -28,43 +29,83 @@ def __init__(self, title='', submitter='', authors='', message='', topic='hermes
+
+    try:
+        response = requests.post(url=submit_url, json=alert, headers=headers)
+        response.raise_for_status()
```
Interesting... I guess I'll just return the response in the exception and hope the HTTP request succeeded and just returned an error response (like a 400)? I added the `raise_for_status` to prevent marking data as shared if it failed, so I'd like to keep that aspect of this at least.

jchate6 replied:
yeah, I think that's a really good improvement. Any way we can get the fact that it's a validation error to the user would be nice.

jchate6 commented on tom_targets/templates/tom_targets/target_share.html:
```diff
@@ -4,16 +4,33 @@
+            {% bootstrap_field form.share_message %}
         </div>
     </div>
     <h4> Include Data </h4>
```
This is true for the target list sharing as well. Turning off the `Include Photometric Data` slider still results in all the data being shared.

jchate6 commented on tom_dataproducts/alertstreams/hermes.py:
```diff
@@ -28,43 +29,83 @@ def __init__(self, title='', submitter='', authors='', message='', topic='hermes
+    stream_base_url = settings.DATA_SHARING['hermes']['BASE_URL']
+    preload_url = stream_base_url + 'api/v0/submit_message/preload/'
+    # You will need your Hermes API key. This can be found on your Hermes profile page.
+    headers = {'Authorization': f"Token {settings.DATA_SHARING['hermes']['HERMES_API_KEY']}"}
```
This Crashes if no API token is found. 
With the removal of the guest user in Hermes, we should remove the Hermes options from the dropdown and open in hermes option if we don't have valid credentials.

jchate6 replied:
At least redirect with an "improperly configured Hermes" error.

jnation3406 commented:
Okay, I've addressed all the issues again. Thank you for your very in-depth testing! Let me know if there are any more issues.

jchate6 reviewed (approved):
I think this is mostly good. I did run into two last bugs but we can spin these off into different issues if you like. 

Specifically, if I don't have a Hermes API key, I get a `KeyError` when trying to open in Hermes. This probably shouldn't happen. If we need the API key, we should handle it gracefully and give the "no API key" error message like in `hermes.py:44`.

The second issue, which I'm guessing is more on the Hermes side is that if I'm not currently logged into Hermes-dev, trying to open a message in Hermes tells me I need to log in, but then doesn't redirect me back to the pre-loaded message.

jnation3406 commented:
> I think this is mostly good. I did run into two last bugs but we can spin these off into different issues if you like.
> 
> Specifically, if I don't have a Hermes API key, I get a `KeyError` when trying to open in Hermes. This probably shouldn't happen. If we need the API key, we should handle it gracefully and give the "no API key" error message like in `hermes.py:44`.

Yeah, it needs an API token since I don't want to allow non-authenticated users to preload (store) data in hermes. I'll see about gateing the open in hermes button around having at API token set at least.

> 
> The second issue, which I'm guessing is more on the Hermes side is that if I'm not currently logged into Hermes-dev, trying to open a message in Hermes tells me I need to log in, but then doesn't redirect me back to the pre-loaded message.

Yes, this is unfortunate but I don't understand the login flow well enough to route it back to where you left from - I'm not even sure if that is possible with the scimma auth login flow... You can go back in your browser once you login since the preload is cached for 15 mintues, or you can click the open in hermes link again to open the page which will work once you've logged in since the login persists.

jnation3406 commented:
Okay, I've hidden the open in hermes button if `HERMES_API_KEY` is not set in DATA_SHARING.
